File: src/ShopNet/Gallery/Get.ts
import dayjs from "dayjs";
import { ShopNet } from "../ShopNet";

/**
 * @author Aloento
 * @since 0.5.0
 * @version 0.1.0
 */
export abstract class GalleryGet extends ShopNet {
  /**
   * @author Aloento
   * @since 0.5.0
   * @version 0.2.1
   */
  public static Categories(): Promise<string[]> {
    return this.WithTimeCache("", "GalleryGetCategories", dayjs().add(1, "m"));
  }

  /**
   * @author Aloento
   * @since 0.5.0
   * @version 0.2.1
   */
  public static async Products(category: string): Promise<[number[], number]> {
    const nums = await this.WithTimeCache<number[]>(category, "GalleryGetProducts", dayjs().add(1, "m"), category);

    return [
      nums,
      4 - (nums.length % 4)
    ];
  }
}

File: TSystems.LoveOTC/Hub/Gallery/Get.cs
namespace TSystems.LoveOTC.Hub;

using Microsoft.EntityFrameworkCore;

internal partial class ShopHub {
    /**
     * <remarks>
     * @author Aloento
     * @since 0.1.0
     * @version 0.2.0
     * </remarks>
     */
    public Task<string[]> GalleryGetCategories() =>
        this.Db.Categories.Select(x => x.Name).ToArrayAsync();

    /**
     * <remarks>
     * @author Aloento
     * @since 0.1.0
     * @version 0.2.0
     * </remarks>
     */
    public Task<uint[]> GalleryGetProducts(string category) =>
        this.Db.Products
            .Where(x => x.Category!.Name == category)
            .Select(x => x.ProductId)
            .ToArrayAsync();
}
